hashset.go

```diff
@@ -0,0 +1,266 @@
+		return false
+	}
+	delete(s.items, key)
+	return true
+}
+
+// RemoveAll will remove each item in items from s.
+//
+// Return true if s was modified (any item was present), false otherwise.
+func (s *HashSet[T, H]) RemoveAll(items []T) bool {
+	modified := false
+	for _, item := range items {
+		if s.Remove(item) {
+			modified = true
+		}
+	}
+	return modified
+}
+
+// RemoveSet will remove each element of o from s.
+//
+// Return true if s was modified (any item of o was present in s), false otherwise.
+func (s *HashSet[T, H]) RemoveSet(o *HashSet[T, H]) bool {
+	modified := false
+	for key := range o.items {
+		if _, exists := s.items[key]; exists {
+			modified = true
+			delete(s.items, key)
+		}
+	}
+	return modified
+}
+
+// Contains returns whether item is present in s.
+func (s *HashSet[T, H]) Contains(item T) bool {
+	_, exists := s.items[item.Hash()]
+	return exists
+}
+
+// ContainsAll returns whether s contains at least every item in items.
+func (s *HashSet[T, H]) ContainsAll(items []T) bool {
+	if len(s.items) < len(items) {
+		return false
+	}
+	for _, item := range items {
+		if !s.Contains(item) {
+			return false
+		}
+	}
+	return true
+}
+
+// ContainsSlice returns whether s contains the same set of of elements
+// that are in items. The elements of items may contain duplicates.
+//
+// If the slice is known to be set-like (no duplicates), EqualSlice provides
+// a more efficient implementation.
+func (s *HashSet[T, H]) ContainsSlice(items []T) bool {
+	return s.Equal(HashSetFrom[T, H](items))
+}
+
+// Subset returns whether o is a subset of s.
+func (s *HashSet[T, H]) Subset(o *HashSet[T, H]) bool {
+	if len(s.items) < len(o.items) {
+		return false
+	}
+	for _, item := range o.items {
+		if !s.Contains(item) {
+			return false
+		}
+	}
+	return true
+}
+
+// Size returns the cardinality of s.
+func (s *HashSet[T, H]) Size() int {
+	return len(s.items)
+}
+
+// Union returns a set that contains all elements of s and o combined.
+func (s *HashSet[T, H]) Union(o *HashSet[T, H]) *HashSet[T, H] {
+	result := NewHashSet[T, H](s.Size())
+	for key, item := range s.items {
+		result.items[key] = item
+	}
+	for key, item := range o.items {
+		result.items[key] = item
+	}
+	return result
+}
+
+// Difference returns a set that contains elements of s that are not in o.
+func (s *HashSet[T, H]) Difference(o *HashSet[T, H]) *HashSet[T, H] {
+	result := NewHashSet[T, H](max(0, s.Size()-o.Size()))
+	for key, item := range s.items {
+		if _, exists := o.items[key]; !exists {
+			result.items[key] = item
+		}
+	}
+	return result
+}
+
+// Intersect returns a set that contains elements that are present in both s and o.
+func (s *HashSet[T, H]) Intersect(o *HashSet[T, H]) *HashSet[T, H] {
+	result := NewHashSet[T, H](0)
+	big, small := s, o
+	if s.Size() < o.Size() {
+		big, small = o, s
+	}
+	for _, item := range small.items {
+		if big.Contains(item) {
+			result.Insert(item)
+		}
+	}
+	return result
+}
+
+// Copy creates a copy of s.
```

Comment:
```suggestion
// Copy creates a shallow copy of s.
```

Since this is intended to be used with pointers-to-structs it's important to clarify this won't copy the structs themselves.

Reply:
good call!